From 80b54257bc2cfb585d500275e731b59a2d56f898 Mon Sep 17 00:00:00 2001
From: Weiwei Wang <weiwei.wang@windriver.com>
Date: Mon, 21 Jun 2010 11:51:05 +0800
Subject: [PATCH] wrhv: x86: jiffies calibration

Guest os has chance to lose hypervisor provided interrupt,
which will cause linux timer precision decline. Here use
wr_vb_status->tick_count to calibrate it.

Signed-off-by: Weiwei Wang <weiwei.wang@windriver.com>
---
 arch/x86/kernel/vbi/wrhv.c |   28 +++++++++++++++++++++++++++-
 1 files changed, 27 insertions(+), 1 deletions(-)

diff --git a/arch/x86/kernel/vbi/wrhv.c b/arch/x86/kernel/vbi/wrhv.c
index ac6c540..9aa51a2 100644
--- a/arch/x86/kernel/vbi/wrhv.c
+++ b/arch/x86/kernel/vbi/wrhv.c
@@ -17,6 +17,7 @@
 #include <linux/pci.h>
 #include <linux/wrhv.h>
 #include <linux/kgdb.h>
+#include <linux/kernel_stat.h>
 #include <vbi/vbi.h>
 #include <asm/setup.h>
 #include <asm/paravirt.h>
@@ -180,8 +181,12 @@ irqreturn_t wrhv_dummy_timer_interrupt(int irq, void *dev_id)
 	return IRQ_HANDLED;
 }
 
+static DEFINE_PER_CPU(u64, mark_offset);
+static DEFINE_PER_CPU(int, mark_first_time) = 1;
+
 irqreturn_t wrhv_timer_interrupt(int irq, void *dev_id)
 {
+	u64 ticks;
 	int cpu = smp_processor_id();
 	struct clock_event_device *evt = &per_cpu(wrhv_clock_events, cpu);
 
@@ -191,7 +196,28 @@ irqreturn_t wrhv_timer_interrupt(int irq, void *dev_id)
 		return IRQ_NONE;
 	}
 
-	evt->event_handler(evt);
+	if (__get_cpu_var(mark_first_time) == 0) {
+		ticks = wr_vb_status->tick_count;
+		ticks -= __get_cpu_var(mark_offset);
+		__get_cpu_var(mark_offset) = wr_vb_status->tick_count;
+		if (ticks > (2*HZ)) {
+			printk(KERN_DEBUG "Time falling behind %lld jiffies\n",
+				ticks);
+			ticks = 1;
+		}
+	} else {
+		ticks = 1;
+		__get_cpu_var(mark_first_time) = 0;
+		__get_cpu_var(mark_offset) = wr_vb_status->tick_count;
+	}
+
+	if (ticks > 1)
+		account_steal_time(jiffies_to_cputime(ticks - 1));
+
+	while (ticks != 0) {
+		evt->event_handler(evt);
+		ticks--;
+	}
 
 	return IRQ_HANDLED;
 }
-- 
1.7.4.1

